src/Temporalio/Worker/WorkflowWorker.cs

```diff
@@ -0,0 +1,254 @@
+                }
+                // Wait on every task to complete (exceptions should never happen)
+                await Task.WhenAll(tasks.Keys).ConfigureAwait(false);
+                var deadlockCount = deadlockedWorkflows.Count;
+                if (deadlockCount > 0)
+                {
+                    logger.LogWarning(
+                        "Worker shutdown with {DeadlockedTaskCount} deadlocked workflow tasks still running",
+                        deadlockCount);
+                }
+            }
+        }
+
+        private async Task HandleActivationAsync(IPayloadCodec? codec, WorkflowActivation act)
+        {
+            WorkflowActivationCompletion comp;
+            RemoveFromCache? removeJob = null;
+
+            // Catch any exception as a completion failure
+            try
+            {
+                // Decode the activation if there is a codec
+                if (codec != null)
+                {
+                    await WorkflowCodecHelper.DecodeAsync(codec, act).ConfigureAwait(false);
+                }
+
+                // Log proto at trace level
+                logger.LogTrace("Received workflow activation: {Activation}", act);
+
+                // We only have to run if there are any non-remove jobs
+                removeJob = act.Jobs.Select(job => job.RemoveFromCache).FirstOrDefault(job => job != null);
+                if (act.Jobs.Count > 1 || removeJob == null)
+                {
+                    // If the activation never completed before, we need to assume continually
+                    // deadlocked and just rethrow the same exception as before
+                    if (deadlockedWorkflows.ContainsKey(act.RunId))
+                    {
+                        throw new InvalidOperationException($"Workflow with ID {act.RunId} deadlocked after {deadlockTimeout}");
+                    }
+
+                    // If the workflow is not yet running, create it. We know that we will only get
+                    // one activation per workflow at a time, so GetOrAdd is safe for our use.
+                    var workflow = runningWorkflows.GetOrAdd(act.RunId, _ => CreateInstance(act));
+
+                    // Activate or timeout with deadlock timeout
+                    // TODO(cretz): Any reason for users to need to customize factory here?
+                    // TODO(cretz): Since deadlocks can cause unreclaimed threads, we are setting
+                    // LongRunning here. That means they may oversubscribe and create a new thread
+                    // just for this, is that ok?
+                    var workflowTask = Task.Factory.StartNew(
+                        () => workflow.Activate(act),
+                        default,
+                        TaskCreationOptions.LongRunning | TaskCreationOptions.DenyChildAttach | TaskCreationOptions.PreferFairness,
+                        TaskScheduler.Current);
+                    var timeoutTask = Task.Delay(deadlockTimeout);
```

Comment:
Create this with a cancel token that we cancel when it didn't timeout. Otherwise this task will complete later unnecessarily (see https://devblogs.microsoft.com/oldnewthing/20220505-00/?p=106585).

I can't find a much cheaper way to timeout a task than create a delayed task for every single task run :-( https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync is only supported in newer .NET versions. Looking [at the source](https://github.com/dotnet/runtime/blob/dfb3d24cec0a227817420de0ae95f45b210e7c25/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L2798-L2816) it basically does what Delay does (timer queue timer) though maybe a couple fewer steps. I will only move to it if microbenchmarks confirm necessity.

I should at least at the above in a comment here.

Reply:
Fixed in next PR